src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepExecution.java

```diff
@@ -0,0 +1,244 @@
+     * Value for SSH_AUTH_SOCK environment variable.
+     */
+    private String socket;
+
+    /**
+     * TODO: Add description.
+     */
+    private List<String> sockets;
+
+    /**
+     * The proxy for the real remote agent that is on the other side of the channel (as the agent needs to
+     * run on a remote machine)
+     */
+    private transient RemoteAgent agent = null;
+
+    @Override
+    public boolean start() throws Exception {
+        StepContext context = getContext();
+        sockets = new ArrayList<String>();
+        initRemoteAgent();
+        context.newBodyInvoker().
+                withContext(EnvironmentExpander.merge(getContext().get(EnvironmentExpander.class), new ExpanderImpl(this))).
+                withCallback(new Callback(this)).withDisplayName(null).start();
+        return false;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (agent != null) {
+            agent.stop();
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
+        }
+        purgeSockets();
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        try {
+            purgeSockets();
+            initRemoteAgent();
+        } catch (IOException io) {
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_CouldNotStartAgent());
+        }
+    }
+
+    private static class Callback extends BodyExecutionCallback {
+
+        private static final long serialVersionUID = 1L;
+
+        private SSHAgentStepExecution execution;
+
+        Callback (SSHAgentStepExecution execution) {
+            this.execution = execution;
+        }
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            try {
+                TaskListener listener = context.get(TaskListener.class);
+                if (execution.getSSHAgent() != null) {
+                    String socket = execution.getSSHAgent().getSocket();
+                    execution.getSSHAgent().stop();
+                    listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped() + " Socket: " + socket);
+                }
+            } catch (Throwable th) {
+                context.onFailure(th);
+            }
+            execution.purgeSockets();
+            context.onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(StepContext context, Throwable t) {
+            try {
+                TaskListener listener = context.get(TaskListener.class);
+                if (execution.getSSHAgent() != null) {
+                    String socket = execution.getSSHAgent().getSocket();
+                    execution.getSSHAgent().stop();
+                    listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped() + " Socket: " + socket);
+                }
+            } catch (Throwable th) {
+                context.onFailure(th);
+            }
+            execution.purgeSockets();
```

Comment:
And then could use `BodyExecutionCallback.TailCall` from https://github.com/jenkinsci/workflow-plugin/pull/168.